infrastructure/src/main/java/org/corfudb/infrastructure/datastore/DataStore.java

```diff
@@ -126,18 +129,31 @@ public synchronized void write(@Nonnull String key, @Nonnull Object value) {
                             Path path = Paths.get(logDirPath, dsFileName);
                             Path tmpPath = Paths.get(logDirPath, dsFileName + ".tmp");
 
-                            String jsonPayload = JsonUtils.parser.toJson(value, value.getClass());
-                            byte[] bytes = jsonPayload.getBytes();
-
-                            ByteBuffer buffer = ByteBuffer.allocate(bytes.length
-                                    + Integer.BYTES);
-                            buffer.putInt(getChecksum(bytes));
-                            buffer.put(bytes);
-                            Files.write(tmpPath, buffer.array(), StandardOpenOption.CREATE,
+                            ByteBuffer dataBuf;
+                            if (value instanceof ByteBuffer) {
+                                dataBuf = ((ByteBuffer) value);
+                            } else {
+                                String jsonPayload = JsonUtils.parser.toJson(value, value.getClass());
+                                dataBuf = ByteBuffer.wrap(jsonPayload.getBytes());
+                            }
```

Comment:
Rather than doing instanceOf checks, why don't you just do overloading?

```
write(@Nonnull String key, @Nonnull Object value) {
  String jsonPayload = JsonUtils.parser.toJson(value, value.getClass());
  dataBuf = ByteBuffer.wrap(jsonPayload.getBytes());
  write(key, dataBuf);
}

write(@Nonnull String key, @Nonnull ByteBuffer value) {
  // Core implementation.
}
```